File: scripts/fill-shopify-colour-families.mjs
import { readFileSync, writeFileSync } from 'fs';

// ── helpers ────────────────────────────────────────────────────────────────
function parseCSV(text) {
  const rows = [];
  let row = [], field = '', inQ = false;
  for (let i = 0; i < text.length; i++) {
    const ch = text[i];
    if (inQ) {
      if (ch === '"' && text[i + 1] === '"') { field += '"'; i++; }
      else if (ch === '"') { inQ = false; }
      else { field += ch; }
    } else {
      if (ch === '"') { inQ = true; }
      else if (ch === ',') { row.push(field); field = ''; }
      else if (ch === '\n') { row.push(field); rows.push(row); row = []; field = ''; }
      else if (ch !== '\r') { field += ch; }
    }
  }
  if (field || row.length) { row.push(field); rows.push(row); }
  return rows;
}

function toCSVField(v) {
  if (v == null) return '';
  const s = String(v);
  return (s.includes(',') || s.includes('"') || s.includes('\n'))
    ? '"' + s.replace(/"/g, '""') + '"'
    : s;
}

function normalize(s) {
  return s.toLowerCase().replace(/\s+/g, ' ').trim();
}

// ── build colour-family lookup from our CSV ────────────────────────────────
// solid-gel-polish-colour-families.csv columns: code, product_name, colour_family
const familyRows = parseCSV(readFileSync('solid-gel-polish-colour-families.csv', 'utf8'));
// key: normalized product_name → colour_family
const familyByName = new Map();
for (const [, productName, colourFamily] of familyRows.slice(1)) {
  if (productName && colourFamily) {
    familyByName.set(normalize(productName), colourFamily);
  }
}

// ── process Shopify export ─────────────────────────────────────────────────
const shopifyRows = parseCSV(readFileSync('products_export_1.csv', 'utf8'));
const header = shopifyRows[0];
// Column B = index 1 (Title), Column K = index 10 (Colour Family)
const titleIdx = 1;
const familyIdx = 10;

let filled = 0, skipped = 0;

const output = [header];
for (const row of shopifyRows.slice(1)) {
  // Ensure row is long enough
  while (row.length <= familyIdx) row.push('');

  const title = row[titleIdx] ?? '';
  const existing = row[familyIdx] ?? '';

  if (!existing && title) {
    const match = familyByName.get(normalize(title));
    if (match) {
      row[familyIdx] = match;
      filled++;
    } else {
      skipped++;
    }
  }
  output.push(row);
}

const outText = output.map(r => r.map(toCSVField).join(',')).join('\n');
writeFileSync('products_export_1_with_families.csv', outText);
console.log(`Done. Filled: ${filled} | No match found: ${skipped}`);

File: scripts/fill-shopify-colour-families.py
import csv
import re
import sys

def to_handle(name):
    """Convert a product name to a Shopify-style handle."""
    s = name.lower()
    s = re.sub(r'[^a-z0-9]+', '-', s)
    s = s.strip('-')
    return s

# ── build lookup: handle → colour_family ──────────────────────────────────
family_by_handle = {}
with open('solid-gel-polish-colour-families.csv', newline='', encoding='utf-8') as f:
    for row in csv.DictReader(f):
        name = row.get('product_name', '').strip()
        family = row.get('colour_family', '').strip()
        if name and family:
            family_by_handle[to_handle(name)] = family

# ── process Shopify export ─────────────────────────────────────────────────
input_file  = 'products_export_1.csv'
output_file = 'products_export_1_with_families.csv'

with open(input_file, newline='', encoding='utf-8') as fin, \
     open(output_file, 'w', newline='', encoding='utf-8') as fout:

    reader = csv.reader(fin)
    writer = csv.writer(fout)

    header = next(reader)
    # Column A = index 0 (Handle), Column K = index 10 (Colour Family)
    handle_idx = 0
    family_idx = 10

    tags_idx = 6  # Column G

    writer.writerow(header)
    filled = 0
    skipped = 0

    for row in reader:
        # Ensure row is wide enough
        while len(row) <= family_idx:
            row.append('')

        handle   = row[handle_idx].strip()
        existing = row[family_idx].strip()

        if not existing and handle:
            match = family_by_handle.get(handle)
            if match:
                row[family_idx] = match
                row[tags_idx] = f'Solid Gel Colours, {match}, Gel Polish'
                filled += 1
            else:
                skipped += 1
        elif existing:
            # Already had a colour family — still set the tag
            row[tags_idx] = f'Solid Gel Colours, {existing}, Gel Polish'

        writer.writerow(row)

print(f"Done. Filled: {filled} | No match found: {skipped}")
print(f"Output: {output_file}")
